### DISCUSSION TITLE
Simple MQTT output of inverter power readings

### DISCUSSION THREAD

**Discussion by bobha43** [SBFspot/SBFspot#645] [Q&A]
I am building a pretty standard home IOT monitoring system based on sensors/mqtt/node-red/ influxDB and grafana. 
The first parts are working well, with
- BME280 sensors (on Pico W microcontrollers / microPython / umqtt) delivering temperature, humidity and atmospheric pressure readings
- the gas and electricity outputs from my smartmeters (RPi 3A+ / Python script / PAHO mqtt client)  
  
I now want to add (only) the power output from my Sunny Boy 3.6 inverter (to compare with the P1 meter readings)  and hope to use SBFSpot installed on a RPI 3B+.   
I can connect to my inverter (ethernet)
I don't need (or want) long term storage (SQL or CSV) on the 3B+, nor do I need all the other data that SBFSpot can deliver. 
I do not plan to send data to PVOutput.
What is the best approach for extracting and publishing the 5-minute power readings as (timestamped) mqtt messages?
Or is SBFSpot overkill for my purposes and would I be better off scraping the power values from my inverter's web interface?

**Comment by sbf-** [comment]
> What is the best approach for extracting and publishing the 5-minute power readings as (timestamped) mqtt messages?

With the current version of SBFspot, the 5min power values can't be send via MQTT. Only the spot values (the power at the time of running SBFspot)

**Comment by bobha43** [comment]
Thanks for your clear reply, even though it is negative for my purposes.
Can you perhaps suggest an alternative to web-scraping  (of which I have no experience) from the inverter's web interface?

**Reply by sbf-** [reply]
> even though it is negative for my purposes

What exactly do you have in mind? Sending the last known 5min value should be possible to add.

**Comment by bobha43** [comment]
Thanks for replying. I simply want to output regular (roughly real-time)
values of the kW delivered by my inverter as mqtt messages, in the same way
as I send kW values (derived from the kWh readings) from my P1 smartmeter.
If I can extract the kW values (in an infinite loop) I can handle the mqtt
part myself using PAHO.

**Reply by sbf-** [reply]
You can run SBFspot every 5 à 10 seconds to publish the actual power value (use `-ad0 -am0 -ae0 -nocsv -nosql` options to disable the things you don't need)
Some time ago, somebody [forked my project](https://github.com/fasihi01/SBFspot) to run it continuously for MQTT purposes

**Comment by bobha43** [comment]
Ok, thanks for that. I'll look into it. Do I understand that I can set up
SBFspot to publish the mqtt message directly and that I
should just organise my Python script to run it regularly?

**Reply by sbf-** [reply]
> Do I understand that I can set up SBFspot to publish the mqtt message directly and that I should just organise my Python script to run it regularly?

Yes, complete the MQTT settings in the config and add the `-mqtt` option to the command line

**Comment by stripwax** [comment]
This kind of recipe might go well into the MQTT doc
https://github.com/SBFspot/SBFspot/wiki/MQTT#quick-reference-for-sbfspot-mqtt
(or perhaps a dedicated more-in-detail page, maybe Quickstart doesn't fit)

**Comment by bobha43** [comment]
Thanks, but (in spite of a lot of searching) I haven't found any
documentation that refers to those options, nor of their use. Where should
I look for that?

Op di 30 mei 2023 om 10:26 schreef SBF ***@***.***>:

**Reply by sbf-** [reply]
SBFspot -? shows you all the options 😉

```
SBFspot [-options]
 -scan               Scan for bluetooth enabled SMA inverters.
 -d#                 Set debug level: 0-5 (0=none, default=2)
 -v#                 Set verbose output level: 0-5 (0=none, default=2)
 -ad#                Set #days for archived daydata: 0-300
                     0=disabled, 1=today (default), ...
 -am#                Set #months for archived monthdata: 0-300
                     0=disabled, 1=current month (default), ...
 -ae#                Set #months for archived events: 0-300
                     0=disabled, 1=current month (default), ...
 -cfg:filename.ext   Set alternative config file
 -finq               Force Inquiry (Inquire inverter also during the night)
 -q                  Quiet (No output)
 -nocsv              Disables CSV export (Overrules CSV_Export in config)
 -nosql              Disables SQL export
 -sp0                Disables Spot.csv export
 -installer          Login as installer
 -password:xxxx      Installer password
 -loadlive           Use predefined settings for manual upload to pvoutput.org
 -startdate:YYYYMMDD Set start date for historic data retrieval
 -settime            Sync inverter time with host time
 -mqtt               Publish spot data to MQTT broker
```

**Comment by bobha43** [comment]
Thanks for your previous help.
I have now got SBFspot delivering  mqtt messages once per minute,
containing only the current power output, as required.
I can read them into Node-red using my mosquitto broker.
The format shown in a debug node is "{"PACTot" : 2500.000 }".
I have tried a lot of ways to convert this in a node-red function node so
that InfluxDB will accept it, but so far without success. I had no problems
earlier with my temperature sensors and my smartmeter outputs.
Do I need to change the SBFspot settings, or can you perhaps suggest how to
reformat the messages in Node-red?
Bob Harris

Op wo 31 mei 2023 om 20:12 schreef SBF ***@***.***>:

**Reply by sbf-** [reply]
Sorry, I'm not familiar with node red...
You can ask at [tweakers.net](https://gathering.tweakers.net/forum/list_messages/1489893) and mention [AUijtdehaag](https://tweakers.net/gallery/192109)
It's a Dutch forum, but I guess that's not a problem for you ;-)

**Comment by bobha43** [comment]
Thanks. In the meantime I realised that SBFspot was outputting JSON code. I
simply inserted a JSON parsing node which outputs a JS object and the
problem went away. I am now able to store the inverter power readings in
InfluxDB and display them as 5 minute averages in Grafana.
SBFspot allows me to select from an enormous range of inverter data and I
may well add more to my system in the future.
Great tool. Thanks for making it available.

**Comment by Infernoken76** [comment]
Hi,

this looks exactly like what I'm looking for...

I have my SBFspot uploading to PVOutput and also via MQTT to my local home assistant.
PVOutput upload is every 5 minutes through crontab, but would like MQTT to be more often...
`
*/5 5-22 * * * /usr/local/bin/sbfspot.3/daydata
`
Above I see that I should be able to do what I like, but unsure what files to edit and what code to add?
Some more guidance please ???

So if I look at the current "daydata" structure is the following and runs through crontab every 5 minutes:
```
#!/bin/bash
#
log=/var/log/sbfspot.3/SB4000-TL20_$(date '+%Y%m%d').log
/usr/local/bin/sbfspot.3/SBFspot -v -ad1 -am0 -ae1 -mqtt -cfg/usr/local/bin/sbfspot.3/SBFspot.cfg $1 $2 $3 $4 $5 &>>$log
```

Do I create a separate file just for MQTT and run more often?  So something like this:
File: mqttdata
```
#!/bin/bash
#
/usr/local/bin/sbfspot.3/SBFspot -v -mqtt
```

Andrun that every minute? 
So:
`
*/1 5-22 * * * /usr/local/bin/sbfspot.3/mqttdata
`

**Comment by Charlotte-br560** [comment]
Sure, here's a concise version:

`import requests
from bs4 import BeautifulSoup
import paho.mqtt.publish as publish
import time

while True:
    response = requests.get('http://your_inverter_ip')
    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')
        power_reading = soup.find(id='power_reading_id').text
        publish.single("your_topic", payload=power_reading, hostname="mqtt_broker_ip")
    time.sleep(300)
`
Replace `'http://your_inverter_ip'`, '`power_reading_id'`, `'your_topic'`, and `'mqtt_broker_ip'` with your specific details. This script fetches the power reading from your inverter's web interface, publishes it as an MQTT message, and repeats it every 5 minutes.
